python/ctsm/subset_data.py

```diff
@@ -0,0 +1,577 @@
+            required=False,
+        )
+        subparser.add_argument(
+            "--datm-syr",
+            help="Start year for creating DATM forcing at single point/region. [default: %("
+                 "default)s]",
+            action="store",
+            dest="datm_syr",
+            required=False,
+            type=int,
+            default=1901,
+        )
+        subparser.add_argument(
+            "--datm-eyr",
+            help="End year for creating DATM forcing at single point/region. "
+                 "[default: %(default)s]",
+            action="store",
+            dest="datm_eyr",
+            required=False,
+            type=int,
+            default=2014,
+        )
+        subparser.add_argument(
+            "--crop",
+            help="Create datasets using the extensive list of prognostic crop types.",
+            action="store_true",
+            dest="crop_flag",
+            required=False,
+        )
+
+        if subparser == pt_parser:
+            parser_name = "single_point"
+        else:
+            parser_name = "regional"
+
+        subparser.add_argument(
+            "--outdir",
+            help="Output directory. \n [default: %(default)s]",
+            action="store",
+            dest="out_dir",
+            type=str,
+            default=os.path.join(os.getcwd(), "subset_data_" + parser_name),
+        )
+        subparser.add_argument(
+            "--user-mods-dir",
+            help="User mods directory.",
+            action="store",
+            dest="user_mods_dir",
+            type=str,
+            default="",
+        )
+        add_logging_args(subparser)
+
+    # -- print help for both subparsers
+    parser.epilog = textwrap.dedent(
+        f"""\
+         {pt_parser.format_help()}
+         {rg_parser.format_help()}
+         """
+    )
+    return parser
+
+def setup_user_mods(user_mods_dir, cesmroot):
+    """
+    Sets up the user mods files and directories
+    """
+    if not os.path.isdir(user_mods_dir):
+        os.mkdir(user_mods_dir)
+
+    nl_clm_base = os.path.join(cesmroot, "cime_config/user_nl_clm")
+    nl_clm = os.path.join(user_mods_dir, "user_nl_clm")
+    with open(nl_clm_base, "r") as basefile, open(nl_clm, "w") as user_file:
+        for line in basefile:
+            user_file.write(line)
+
+    nl_datm_base = os.path.join(cesmroot, "components/cdeps/datm/cime_config"
+                                          "/user_nl_datm_streams")
+    nl_datm = os.path.join(user_mods_dir, "user_nl_datm_streams")
+    with open(nl_datm_base, "r") as base_file, open(nl_datm, 'w') as user_file:
+        for line in base_file:
+            user_file.write(line)
+
+def determine_num_pft (crop):
+    """
+    A simple function to determine the number of pfts.
+
+    Args:
+        crop (bool): crop flag denoting if we are using crop
+
+    Raises:
+
+    Returns:
+        num_pft (int) : number of pfts for surface dataset
+    """
+    if crop:
+        num_pft = "78"
```

Comment:
[ ] Use the constants from the top.

Reply:
@negin513  and I went over this, and we need to revisit. The constants are in the single point module and not here, so can't be used. So they would have to be redefined or #1621 also done. In this context they are being used to capture the right filenames, but the values do relate to the size of natural PFT's + the two generic crops, and total number of all possible PFT's. So if the numbers were changed in the single point module (because they were changed in CTSM) they would also have to change here in subset_data. Which is why putting them in a file for constants might be a good thing to do.

Reply:
@ekluzek and I discussed about this a bit in our meeting. Essentially this is different from the constants (NAT_PFT and MAX_PFT) introduced. 

NAT_PFT (15) and MAX_PFT (78) represent the dimension of the original surface dataset file. 
However, the values for num_pft are only used for the names of the files and the rest of the file names are hard-coded anyways. 

I am okay with changing these, but please note that I think there might be potentially repercussions:

In single_point_case.py we introduced NAT_PFT (which is the number of natural pfts) as 15 and MAX_PFT as 78. Here for the file names we are pointing to files that has 16 and 78 in their file names.

Are we going to use these numbers (PFT numbers) similarly in future? 
**For example the files with 15 PFTS is named as 16 (NAT_PFT +1) but the file with 78 PFT is named as 78 .** 
**For example, in future if the number of MAX_PFTs are increased then does our file names always point to MAX_PFT or MAX_PFT+1?** This is especially important for me because for one file we are using 16 (NAT_PFT+1) but for the other file we are using 78 PFTS and not 78+1 (MAX_PFT+1). 

I think it might be potentially dangerous or error-prone to tie the filenames here.  Especially because we are not following consistent naming for our files. For example the 15 PFT is named as 16 but the 78 PFTs is named as 78 (instead of 79). 
This is quite an easy change, I just want to make sure it does not cause confusions in future.

Reply:
I agree that this is confusing. It seems that `num_pft` is only used in determining the file names and functioning as a string rather than a variable, so I think it's probably ok to treat this as a string (e.g. specify the number) rather than use variables in this particular case.

Reply:
Yes, since this is hard tied to the filenames, and it's a bit confusing how this relates to the other constants, we thought it best to just hardcode it like this for now. The filenames by nature aren't going to change very often.